pkg/networkplugin-syncer/handlers/ovn/infrastructure.go

```diff
@@ -42,21 +46,73 @@ func (ovn *SyncHandler) ensureSubmarinerInfra() error {
 	// At this point, we are missing the ovn_cluster_router policies and the
 	// local-to-remote & remote-to-local routes in submariner_router, that
 	// depends on endpoint details that we will receive via events
-
 	return nil
 }
 
+// ensureSubmarinerGatewayLocalNetSwitch creates the OVN submariner_join switch which
+// connects the ovn_cluster_router to the submariner_router
 func (ovn *SyncHandler) ensureSubmarinerJoinSwitch() error {
-	klog.Infof("Ensuring %q switch", submarinerDownstreamSwitch)
-	lsCmd, err := ovn.nbdb.LSAdd(submarinerDownstreamSwitch)
+	return ovn.ensureSwitch(submarinerDownstreamSwitch)
+}
+
+// ensureSubmarinerGatewayLocalNetSwitch creates the OVN submariner_gateway switch and
+// virtual ports necessary to make this switch connected to br-submariner on each worker
+// node. This switch is used by the submariner gateway to connect to gateway node and remote
+// clusters.
+func (ovn *SyncHandler) ensureSubmarinerGatewayLocalNetSwitch() error {
+	err := ovn.ensureSwitch(submarinerUpstreamSwitch)
+	if err != nil {
+		return err
+	}
+
+	lspCmd, err := ovn.nbdb.LSPAdd(submarinerUpstreamSwitch, submarinerUpstreamLocalnetPort)
+	if err == nil {
+		err = ovn.nbdb.Execute(lspCmd)
+	}
+
+	if err != nil && !errors.Is(err, goovn.ErrorExist) {
+		return errors.Wrapf(err, "error creating Submariner localnet port %q", submarinerUpstreamLocalnetPort)
+	}
+
+	lspTypeCmd, err := ovn.nbdb.LSPSetType(submarinerUpstreamLocalnetPort, "localnet")
+	if err == nil {
+		err = ovn.nbdb.Execute(lspTypeCmd)
+	}
+
+	if err != nil {
+		return errors.Wrapf(err, "error setting Submariner localnet port %q type", submarinerUpstreamLocalnetPort)
+	}
+
+	lspSetAddressesCmd, err := ovn.nbdb.LSPSetAddress(submarinerUpstreamLocalnetPort, "unknown")
+	if err == nil {
+		err = ovn.nbdb.Execute(lspSetAddressesCmd)
+	}
+
+	if err != nil {
+		return errors.Wrapf(err, "error setting Submariner localnet port %q address", submarinerUpstreamLocalnetPort)
+	}
+
+	lspSetOptionsCmd, err := ovn.nbdb.LSPSetOptions(
+		submarinerUpstreamLocalnetPort, map[string]string{"network_name": SubmarinerUpstreamLocalnet})
+
+	if err == nil {
+		err = ovn.nbdb.Execute(lspSetOptionsCmd)
```

Comment:
If err is nil, we still return error at L102

Reply:
No, the wrap implementation returns nil if the error is nil :)

```
// Wrap returns an error annotating err with a stack trace
// at the point Wrap is called, and the supplied message.
// If err is nil, Wrap returns nil.
func Wrap(err error, message string) error {
	if err == nil {
		return nil
	}
	err = &withMessage{
		cause: err,
		msg:   message,
	}
	return &withStack{
		err,
		callers(),
	}
}
```

Reply:
Anyway, let me tune the implementation to make it less confusing. I must admit I was also confused by that once

Reply:
tuned that bit to make it more obvious.